Realm/RLMObservation.mm

```diff
@@ -550,3 +552,70 @@ void RLMDidChange(std::vector<realm::BindingContext::ObserverState> const& obser
         static_cast<RLMObservationInfo *>(info)->didChange(RLMInvalidatedKey);
     }
 }
+
+KeyPath KeyPathFromString(RLMRealm *realm,
+                          RLMSchema *schema,
+                          RLMClassInfo *info,
+                          RLMObjectSchema *rlmObjectSchema,
+                          NSString *keyPath) {
+    RLMProperty *property;
+    KeyPath keyPairs;
+
+    NSUInteger start = 0, length = keyPath.length, end = NSNotFound;
+    do {
```

Comment:
This block would still need some way to catch a bogus string, like ["array.nonExistantProperty", "nonExistantProperty"]. 

```
//...
property = rlmObjectSchema[component];
RLMPrecondition(property, @"Invalid property name",
                        @"Property '%@' not found in object of type '%@'",
                        component, rlmObjectSchema.className);
TableKey tk = info->objectSchema->table_key;
ColKey ck;
if (property.type == RLMPropertyTypeObject) {
//...
```

Do you think there are better ways to check this? I'm asking because the way I expose RLMPrecondition smells bad.

Reply:
Removing the do/while does make this infinitely easier to read.

Reply:
This code originally used `componentsSeparatedByString:` and I switched it to this more complicated form as [part of some optimizing work years ago](https://github.com/realm/realm-cocoa/pull/1956). I assume that means it showed up as taking enough time in a profile to be worth looking at, but the net change numbers in the PR are the cumulatively result of a bunch of changes and I didn't record the impact of each individual thing. Might be worth at least rerunning the benchmark to see if it's still a measurable difference?

Reply:
Kind of surprising since this isn't really a hot path, but yeah, you must have done it for some reason. @ericjordanmossman do you think you'd be able to measure this?

Reply:
Also having precondition checks is fine.

Reply:
For something like `[[Foo objectsInRealm:realm where:@"prop > 1"] count]` where running the query is very fast, the conversion from NSPredicate to a core Query was once a substantial portion of the runtime. It has been something like 6 years since I last looked into this, though, so it might be totally different or totally irrelevant by now. They've done a lot to improve the performance of obj-c code in general since then and a lot of incidental optimizations have gone from 20% speedups to more like 5%.

The thing to test here is just to run the testQueryConstruction performance test on master and then again on this branch to see if it's even changed.